File: src/linearize.py
import numpy as np
from math import log, exp
from .utils import *

class Polyfit:
    '''
    Polyfit model;
    see Linearization.py for details;
    '''
    def __init__(self, s, d, deg):
        p = np.polyfit(s, d, deg)
        self.p = np.poly1d(p)
    
    def __call__(self, inp):
        return self.p(inp)

class Logpolyfit:
    '''
    Logpolyfit model;
    see Linearization.py for details;
    '''
    def __init__(self, s, d, deg):
        mask = (s>0) & (d>0)
        s = s[mask]
        d = d[mask]
        self.p = np.poly1d(np.polyfit(np.log(s), np.log(d), deg))
    
    def __call__(self, inp):
        mask = inp>0
        output = inp.copy()
        output[mask] = np.exp(self.p(np.log(output[mask])))
        output[~mask] = 0
        return output


class Linear:
    '''linearization base'''
    def __init__(self, *args):
        '''get args'''
        pass

    def calc(self):
        '''calculate parameters'''
        pass

    def linearize(self, inp):
        '''inference'''
        return inp
    
    def value(self):
        '''evaluate linearization model'''
        pass

class Linear_identity(Linear):
    '''make no change'''
    pass

class Linear_gamma(Linear):
    '''
    gamma correction;
    see Linearization.py for details;
    '''
    def __init__(self, gamma, *_):
        self.gamma = gamma

    def linearize(self, inp):
        return gamma_correction(inp, self.gamma)

class Linear_color_polyfit(Linear):
    '''
    polynomial fitting channels respectively;
    see Linearization.py for details;
    '''    
    method = Polyfit
    def __init__(self, _, deg, src, dst, mask, cs):
        self.deg = deg
        # mask = saturate(src, *saturated_threshold)
        self.src, self.dst = src[mask], dst.to(cs.l).colors[mask]
        self.calc()

    def calc(self):
        '''
        monotonically increase is not guaranteed;
        see Linearization.py for more details;
        '''
        rs, gs, bs = self.src[..., 0], self.src[..., 1], self.src[..., 2]
        rd, gd, bd = self.dst[..., 0], self.dst[..., 1], self.dst[..., 2]

        self.pr = self.method(rs, rd, self.deg)
        self.pg = self.method(gs, gd, self.deg)
        self.pb = self.method(bs, bd, self.deg)

    def linearize(self, inp):
        r, g, b = inp[..., 0], inp[..., 1], inp[..., 2]
        return np.stack([self.pr(r), self.pg(g), self.pb(b)], axis=-1)

class Linear_color_logpolyfit(Linear_color_polyfit):
    '''
    logarithmic polynomial fitting channels respectively;
    see Linearization.py for details;
    '''      
    method = Logpolyfit

class Linear_gray_polyfit(Linear):
    '''
    grayscale polynomial fitting;
    see Linearization.py for details;
    '''      
    method = Polyfit
    def __init__(self, _, deg, src, dst, mask, cs):
        self.deg = deg
        dst.get_gray()
        mask = mask & dst.grays

        # the grayscale function is approximate for src is in relative color space;
        # see Linearization.py for more details;
        self.src, self.dst = rgb2gray(src[mask]), dst.toGray(cs.io)[mask]
        self.calc()

    def calc(self):
        '''
        monotonically increase is not guaranteed;
        see Linearization.py for more details;
        '''        
        self.p = self.method(self.src, self.dst, self.deg)
        # self.p = np.poly1d(p)
        # print('p', self.p)

    def linearize(self, inp):
        return self.p(inp)

class Linear_gray_logpolyfit(Linear_gray_polyfit):
    '''
    grayscale logarithmic polynomial fitting;
    see Linearization.py for details;
    '''      
    method = Logpolyfit

File: cpp/linearize.cpp
#include "linearize.h"

Mat Linear::linearize(Mat inp)
{
    return inp;
}
Mat Linear::mask_copyto(Mat src, Mat mask) {
    Mat src_(countNonZero(mask), 1, src.type());
    int countone = 0;
    for (int i = 0; i < mask.rows; i++) {
        if (mask.at<double>(i, 0)) {
            for (int c = 0; c < src.channels(); c++) {
                src_.at<Vec3d>(countone, 0)[c] = src.at<Vec3d>(i, 0)[c];
            }
            countone++;
        };
    }
    return src_;
}
Mat Linear::mask_copyto_gray(Mat src, Mat mask) {
    Mat src_(countNonZero(mask), 1, src.type());
    int countone = 0;
    for (int i = 0; i < mask.rows; i++) {
        if (mask.at<char>(i, 0)) {
            src_.at<double>(countone, 0) = src.at<double>(i, 0);
            countone++;
        };
    }
    return src_;
}

Linear_gamma::Linear_gamma(float gamma_, int deg, Mat src, ColorCheckerMetric cc, vector<double> saturated_threshold)
{
    gamma = gamma_;
}

Mat Linear_gamma::linearize(Mat inp) {
    return gamma_correction(inp, gamma);
}

Linear_color_polyfit::Linear_color_polyfit(float gamma, int deg, Mat src, ColorCheckerMetric cc, vector<double> saturated_threshold) {
    Mat mask = saturate(src, saturated_threshold[0], saturated_threshold[1]);
    Mat src_(countNonZero(mask), 1 ,src.type());
    Mat dst_(countNonZero(mask),1, cc.rgbl.type());
    this->deg = deg;
  
    this->src = mask_copyto(src, mask);
    this->dst = mask_copyto(cc.rgbl, mask);
    calc();
}


void Linear_color_polyfit::calc(void)
{
    Mat sChannels[3];
    Mat dChannels[3];  
    split(this->src, sChannels);
    split(this->dst, dChannels);
    Mat rs = sChannels[0];
    Mat gs = sChannels[1];
    Mat bs = sChannels[2];
    Mat rd = dChannels[0];
    Mat gd = dChannels[1];
    Mat bd = dChannels[2];
    
    pr = polyfit(rs, rd, deg);
    pg = polyfit(gs, gd, deg);
    pb = polyfit(bs, bd, deg);
}


Mat Linear_color_polyfit::linearize(Mat inp)
{
    Mat inpChannels[3];

    split(inp, inpChannels);
    vector<Mat> channel;
    Mat res;

    channel.push_back(poly1d(inpChannels[0], pr, deg));
    channel.push_back(poly1d(inpChannels[1], pg, deg));
    channel.push_back(poly1d(inpChannels[2], pb, deg));
    merge(channel, res);
    return res;
}

Linear_color_logpolyfit::Linear_color_logpolyfit(float gamma, int deg, Mat src, ColorCheckerMetric cc, vector<double> saturated_threshold) {
    Mat mask = saturate(src, saturated_threshold[0], saturated_threshold[1]);
    Mat src_(countNonZero(mask), 1, src.type());
    Mat dst_(countNonZero(mask), 1, cc.rgbl.type());
    this->deg = deg;
    this->src = mask_copyto(src, mask);
    this->dst = mask_copyto(cc.rgbl, mask);
    calc();
}

void Linear_color_logpolyfit::calc(void)
{
    Mat sChannels[3];
    Mat dChannels[3];
    split(src, sChannels);
    split(dst, dChannels);   
    Mat rs = sChannels[0];
    Mat gs = sChannels[1];
    Mat bs = sChannels[2];
    Mat rd = dChannels[0];
    Mat gd = dChannels[1];
    Mat bd = dChannels[2];
    pr = _polyfit(rs, rd, deg);
    pg = _polyfit(gs, gd, deg);
    pb = _polyfit(bs, bd, deg);
}


Mat Linear_color_logpolyfit::linearize(Mat inp)
{
    Mat channels[3];
    split(inp, channels);
    Mat r = channels[0];
    Mat g = channels[1];
    Mat b = channels[2];
    vector<Mat> channel;
    Mat res;
    channel.push_back(_lin(pr, r,deg));
    channel.push_back(_lin(pg, g,deg));
    channel.push_back(_lin(pb, b,deg));
    merge(channel, res);
    return res;

}


Linear_gray_polyfit::Linear_gray_polyfit(float gamma, int deg, Mat src, ColorCheckerMetric cc, vector<double> saturated_threshold) {
    Mat mask = saturate(src, saturated_threshold[0], saturated_threshold[1])& ~cc.white_mask;
    Mat src_(countNonZero(mask), 1, src.type());
    Mat dst_(countNonZero(mask), 1, cc.grayl.type());
    this->deg = deg;
    Mat src_gray = mask_copyto(src, mask);
    this->src = rgb2gray(src_gray);
    this->dst = mask_copyto(cc.grayl, mask);

    calc();
}


void Linear_gray_polyfit::calc(void)
{
    this-> p = polyfit(src, dst, deg);
}


Mat Linear_gray_polyfit::linearize(Mat inp)
{
    Mat inpChannels[3];
    split(inp, inpChannels);
    vector<Mat> channel;
    Mat res;
    channel.push_back(poly1d(inpChannels[0], p, deg));
    channel.push_back(poly1d(inpChannels[1], p, deg));
    channel.push_back(poly1d(inpChannels[2], p, deg));
    merge(channel, res);
    return res;
}


Linear_gray_logpolyfit::Linear_gray_logpolyfit(float gamma, int deg, Mat src, ColorCheckerMetric cc, vector<double> saturated_threshold) {
    Mat mask = saturate(src, saturated_threshold[0], saturated_threshold[1]) & ~cc.white_mask;
    Mat src_(countNonZero(mask), 1, src.type());
    Mat dst_(countNonZero(mask), 1, cc.grayl.type());
    this->deg = deg;

    Mat src_gray = mask_copyto(src, mask);
    this->src = rgb2gray(src_gray);
    
    this->dst = mask_copyto(cc.grayl, mask);
    calc();
}


void Linear_gray_logpolyfit::calc(void)
{
    this->p = _polyfit(src, dst, deg);
}


Mat Linear_gray_logpolyfit::linearize(Mat inp)
{
    Mat inpChannels[3];
    split(inp, inpChannels);
    vector<Mat> channel;
    Mat res;

    channel.push_back(_lin( p, inpChannels[0], deg));
    channel.push_back(_lin( p, inpChannels[1], deg));
    channel.push_back(_lin( p, inpChannels[2], deg));
    merge(channel, res);
    return res;
}


Mat Linear::_polyfit(Mat src, Mat dst, int deg) {
    
    Mat mask_ = (src > 0) & (dst > 0);
    Mat src_, dst_;
    src_ = mask_copyto_gray(src, mask_);
    dst_ = mask_copyto_gray(dst, mask_);
    Mat s, d;
    log(src_, s);
    log(dst_, d);
    Mat res = polyfit(s, d, deg);
    return res;
}


Mat Linear::_lin(Mat p, Mat x, int deg) {
    
    Mat mask_ = x >= 0;
    Mat y;
    log(x, y);
    y = poly1d(y, p, deg);
    Mat y_;
    exp(y, y_);
    Mat res;
    y_.copyTo(res, mask_);
    return res;
}


Mat Linear::polyfit(Mat src_x, Mat src_y, int order) {
    int npoints = src_x.checkVector(1);
    int nypoints = src_y.checkVector(1);
    Mat_<double> srcX(src_x), srcY(src_y);
    Mat_<double> A = Mat_<double>::ones(npoints, order + 1);
    for (int y = 0; y < npoints; ++y)
    {
        for (int x = 1; x < A.cols; ++x)
        {         
            A.at<double>(y, x) = srcX.at<double>(y) * A.at<double>(y, x - 1);           
        }
    }
    Mat w;
    cv::solve(A, srcY, w, DECOMP_SVD);
    return w;
}

Mat Linear::poly1d(Mat src, Mat w, int deg) {
    Mat res_polyfit(src.size(),src.type());
    for (int i = 0; i < src.rows; i++) {
        for (int j = 0; j < src.cols; j++) {
                double res = 0;
                for (int d = 0; d <=deg; d++) {
                    res += pow(src.at<double>(i, j), d) * w.at<double>( d,0);
                    res_polyfit.at<double>(i, j) = res;
            }        
        }
    }  
    return res_polyfit;
}

Linear* get_linear(string linear) {
    Linear* p = new Linear;
    if (linear == "Linear") {
        p = new Linear;
    }
    else if (linear == "Linear_identity") {
        p = new Linear_identity;
    }
    else if (linear == "Linear_gamma") {
        p = new Linear_gamma;
    }
    else if (linear == "Linear_color_polyfit") {
        p = new Linear_color_polyfit;
    }
    else if (linear == "Linear_color_logpolyfit") {
        p = new Linear_color_logpolyfit;
    }
    else if (linear == "Linear_gray_polyfit") {
        p = new Linear_gray_polyfit;
    }
    else if (linear == "Linear_gray_logpolyfit") {
        p = new Linear_gray_logpolyfit;
    }
    return p;
}
